<h1>Linear Algebra (CpE210A)

<h3>Midterms Project

Coded and submitted by: <br>
<i>Parco, Jovian Asher G.<br>
    201911101 <br>
   58010 </i>

Directions
This Jupyter Notebook will serve as your base code for your Midterm Project. You must further format and provide complete discussion on the given topic. 
- Provide all necessary explanations for specific code blocks. 
- Provide illustrations for key results.
- Observe clean code (intuitive variable names, proper commenting, proper code spacing)
- Provide a summary discussion at the end

Failure to use this format or failure to update the document will be given a deduction equivalent to 50% of the original score. 

### Case

Bebang is back to consult you about her business. Furthering her data analytics initiative she asks you for help to compute some relevant data. Now she is asking you to compute and visualize her sales and costs for the past year. She has given you the datasets attached to her request.

### Problem

State and explain Bebang's problem here and provide the deliverables.

# Proof of Concept

Now that you have a grasp on the requirements we need to start with making a program to prove that her problem is solvable. As a Linear Algebra student, we will be focusin on applying vector operations to meet her needs. First, we need to import her data. We will use the `pandas` library for this. For more information you can look into their documentation [here](https://pandas.pydata.org/).

In [1]:
import pandas as pd

## Adding variable with the name of the file being access
file_prices = "bebang prices.csv" 
file_sales = "bebang sales.csv"

## Importing to Python using Panda's library
df_prices = pd.read_csv(file_prices, index_col=0) # Specifying the row name with a blank value  
df_sales = pd.read_csv(file_sales)


## Printing bebang Prices.csv
df_prices

red velvet  oreo  sugar glazed  almond honey  matcha  \
price per piece          35    25            20            35      50   
cost per piece           25    15            10            20      35   

                 strawberry cream  brown sugar boba  furits and nuts  \
price per piece                40                40               45   
cost per piece                 31                30               25   

                 dark chocolate  chocolate and orange  choco mint  \
price per piece              30                    40          45   
cost per piece               15                    25          32   

                 choco butter naught  
price per piece                   45  
cost per piece                    31

In [41]:
## Printing bebang sales.csv
df_sales

flavor  january  february  march  april  may  june  july  \
0             red velvet      781       524    364    157  894    56    89   
1                   oreo      210       324    344     32  123     1     3   
2           sugar glazed      599       674    854     14  467    78    36   
3           almond honey      124       112    114     23  198    12    32   
4                 matcha      645       498    885     98  177    98    43   
5       strawberry cream      210       372    369     78  163    26    45   
6       brown sugar boba      981       380    154     14  147    15    98   
7        furits and nuts      478       462    354     49   89     0     1   
8         dark chocolate      598       203     36     96  789     4     5   
9   chocolate and orange      324       123     48     32   65    74    78   
10            choco mint      458       412     78     78   78     7     3   
11   choco butter naught      778       586    645    100  987    63    36   

    august  september  october  november  december  total  
0       41         23       78       231       899   4137  
1        0          1        6       312       548   1904  
2       48         75       98       781      1602   5326  
3       41         23       36       300       566   1581  
4       96         26       71       549       897   4083  
5       22         54       41       207       456   2043  
6       33         23       36       552       633   3066  
7        0          1        8       335       214   1991  
8        0          8        0       789       458   2986  
9       45         89      100       632       231   1841  
10       8          2        1       323       564   2012  
11      89         78       99       784      2310   6555

## Part 1: Monthly Sales

In [4]:
import numpy as np

sales_mat = np.array(df_sales.set_index('flavor'))  # Vecotorize data
prices_mat = np.array(df_prices)[0].reshape((12,1)) # Vectorize data with 12 row and 1 column
costs_mat = np.array(df_prices)[1].reshape((12,1))  # Vectorize data with 12 row and 1 column


## Printing shapes of the vector
print(sales_mat.shape)
print(prices_mat.shape)
print(costs_mat.shape)

# Expected Output
# (12, 12)
# (12, 1)
# (12, 1)

(12, 12)
(12, 1)
(12, 1)


#### Formulas

Take note that the fomula for revenue is: <br>
$revenue = sales * price $ <br>
In this case, think that revenue, sales, and price are vectors instead of individual values <br>
The formula of cost per item sold is: <br>
$cost_{sold} = sales * cost$ <br>
The formula for profit is: <br>
$profit = revenue - cost_{sold}$ <br>
Solving for the monthly profit will be the sum of all profits made on that month.

In [5]:
## Function that Computes Monthly Sales
def monthly_sales(price, cost, sales):                                                   # Function that Inputs Three Parameters
    monthly_revenue = np.multiply(sales,price)                                           # Solving monthly revenue
    monthly_costs = np.multiply(sales,cost)                                              # Solving monthly cost
    monthly_profits =monthly_revenue - monthly_costs                                     # Solving monthly profits
    return monthly_revenue.flatten(), monthly_costs.flatten(), monthly_profits.flatten() # Outputs as one dimensional array

In [6]:
## Calling monthly_sales() Function with Three Argumments
month_rev, month_cost, month_profit = monthly_sales(prices_mat, costs_mat, sales_mat)

## print the values
print(f"Monthly Revenue:\n {month_rev}\n")
print(f"Monthly Cost:\n {month_cost}\n")
print(f"Monthly Profits:\n {month_profit}")

Monthly Revenue:
 [ 27335  18340  12740   5495  31290   1960   3115   1435    805   2730
   8085  31465   5250   8100   8600    800   3075     25     75      0
     25    150   7800  13700  11980  13480  17080    280   9340   1560
    720    960   1500   1960  15620  32040   4340   3920   3990    805
   6930    420   1120   1435    805   1260  10500  19810  32250  24900
  44250   4900   8850   4900   2150   4800   1300   3550  27450  44850
   8400  14880  14760   3120   6520   1040   1800    880   2160   1640
   8280  18240  39240  15200   6160    560   5880    600   3920   1320
    920   1440  22080  25320  21510  20790  15930   2205   4005      0
     45      0     45    360  15075   9630  17940   6090   1080   2880
  23670    120    150      0    240      0  23670  13740  12960   4920
   1920   1280   2600   2960   3120   1800   3560   4000  25280   9240
  20610  18540   3510   3510   3510    315    135    360     90     45
  14535  25380  35010  26370  29025   4500  44415   2835   

## Part 2: Flavor Sales

In [7]:
## Function that Computes Yearly Sales with Respect to Flavors
def flavor_sales(price, cost, sales):                 # Function that Inputs Three Parameters
    flavor_revenue = np.dot(sales,price)              # solving revenue 
    flavor_costs = np.dot(sales,cost)                 # solving costs
    flavor_profits = flavor_revenue - flavor_costs    # solving profits
    return flavor_profits.flatten()                   # Outputs as one dimensional array


In [43]:
## Calling flavor_profit() Function with Three Argumments
flavor_profit = flavor_sales(prices_mat, costs_mat, sales_mat)

## print the values
print(f"Profits by flavor in a year:\n{flavor_profit}")

Profits by flavor in a year:
[51773 22977 65683 20772 48787 24749 36263 22506 38520 23036 24213 84289]


In [88]:
## Creating New Column of the Sum of Monthly Sales
df_sales["total"]= df_sales["january"] + df_sales["february"] + df_sales["march"] + df_sales["april"] + df_sales["may"] + df_sales["june"] + df_sales["july"] + df_sales["august"] + df_sales["september"] + df_sales["october"] + df_sales["november"] + df_sales["december"] 

##Sorting Values of
sorted_sales_flavor = df_sales.sort_values(["total"], ascending = False, ignore_index=True)

##Identifying Top & Bottom Values
best_3_flavors = sorted_sales_flavor[["flavor","total"]].head(3)
worst_3_flavors = sorted_sales_flavor[["flavor","total"]].tail(3)

## Removing Index
worst_3_flavors = worst_3_flavors.to_string(index=False)
best_3_flavors = best_3_flavors.to_string(index=False)

## Printing Results
print("Top 3 flavors:\n",best_3_flavors,"\n")
print("Worst 3 flavors:\n",worst_3_flavors)



Top 3 flavors:
               flavor  total
 choco butter naught   6555
        sugar glazed   5326
          red velvet   4137 

Worst 3 flavors:
                flavor  total
                 oreo   1904
 chocolate and orange   1841
         almond honey   1581


## Part 3: Visualizing the Data (Optional for +40%)
You can try to visualize the data in the most comprehensible chart that you can use.

In [10]:
import matplotlib.pyplot as plt
import matplotlib
import seaborn as sns

%matplotlib inline

#### Entire Dataset

In [ ]:
None

None Monthly Sales

In [90]:
None

#### Flavor Sales

In [91]:
None

## Part 4: Business Recommendation and Conclusion

The researcher concludes that the worst 3 product is Oreo with a total sales of 1904, chocolate and orange with a total sales of 1841, and almond honey with a  total sales of 1581 recommends to remove it in the shop or to sell it by seasons.
<BR>
The researcher also recommends pricing the top-selling product Red Velvet and Oreo for about 10 pesos more, and  Choco Butter Naught increase for 1 peso, which increases the overall profit.